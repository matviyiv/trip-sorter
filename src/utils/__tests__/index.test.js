import data from './deals.data.json';
import {
  findRoute,
  getAllDestinations,
  durationToMin,
  getRealCost,
  durationToString,
} from '../index';

describe('utils.index', () => {
  describe('findRoute', () => {
    let deals;

    beforeEach(() => {
      deals = data.deals;
    });

    it('return direct route', () => {
      expect(findRoute(deals, 'London', 'Paris', 'cheapest')).toMatchSnapshot();
    });

    it('fastest and cheapest should be diffrent', () => {
      const cheapest = findRoute(deals, 'London', 'Paris', 'cheapest');
      const fastest = findRoute(deals, 'London', 'Paris', 'fastest');
      expect(cheapest).not.toEqual(fastest);
    });

    xit('route with one stop over', () => {
      expect(findRoute(deals, 'London', 'Madrid', 'cheapest')).toMatchSnapshot();
    });

    xit('route with three stop overs', () => {
      expect(findRoute(deals, 'London', 'Stockholm', 'fastest')).toMatchSnapshot();
    });

    it('route with 5 stops', () => {
      expect(findRoute(deals, 'London', 'Kyiv', 'cheapest')).toMatchSnapshot();
    });

    xit('should not arrive to same city twice', () => {
      expect(findRoute(deals, 'London', 'Geneva', 'cheapest')).toMatchSnapshot();
    });

    it('reversed route should also work', () => {
      expect(findRoute(deals, 'Kyiv', 'London', 'cheapest')).toMatchSnapshot();
    });
  });

  describe('getAllDestinations', () => {
    let deals;

    beforeEach(() => {
      deals = data.deals;
    });

    it('should return departureList and arrivalList', () => {
      expect(getAllDestinations(deals)).toMatchSnapshot();
    });
  });

  describe('durationToMin', () => {
    it('convert duration object with one hour', () => {
      expect(durationToMin({h: '01', m: '00'})).toBe(60);
    });

    it('convert duration object with 3 hours and 45 min', () => {
      expect(durationToMin({h: '03', m: '45'})).toBe(225);
    });

    it('fail to convert string values into integer', () => {
      expect(durationToMin({h: 'xx', m: 'yy'})).toEqual(NaN);
    });

    it('fail without parameter', () => {
      expect(durationToMin(undefined)).toEqual(NaN);
    });
  });

  describe('getRealCost', () => {
    it('cost without discount', () => {
      expect(getRealCost({cost: 100, discount: 0})).toBe(100);
    });

    it('cost with 50% discount', () => {
      expect(getRealCost({cost: 100, discount: 50})).toBe(50);
    });
  });

  describe('durationToString', () => {
    it('1h0', () => {
      expect(durationToString(60)).toBe('1h0');
    });
    it('1h40', () => {
      expect(durationToString(100)).toBe('1h40');
    });
  });
});
