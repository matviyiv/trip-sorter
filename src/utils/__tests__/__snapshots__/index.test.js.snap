// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils.index findRoute return direct route 1`] = `
Array [
  "BLP0530",
]
`;

exports[`utils.index findRoute reversed route should also work 1`] = `
Array [
  "BKM0500",
  "BMS0600",
  "BSW0545",
  "BWP0615",
  "BPB0700",
  "BAI0530",
  "BIA0515",
  "BAR0600",
  "BRG0700",
  "BGB0515",
  "BBA0515",
  "BAL0545",
]
`;

exports[`utils.index findRoute route with 5 stops 1`] = `
Array [
  "BLA0745",
  "BAW0515",
  "BWS0515",
  "BSM0530",
  "BMK0700",
]
`;

exports[`utils.index findRoute route with one stop over 1`] = `
Array [
  "BLP0530",
  "BPM0645",
]
`;

exports[`utils.index findRoute route with three stop overs 1`] = `
Array [
  "CLA0445",
  "CAW0445",
  "TWS0515",
]
`;

exports[`utils.index findRoute should not arrive to same city twice 1`] = `
Array [
  "BLP0530",
  "BPB0630",
  "BBG0715",
]
`;

exports[`utils.index getAllDestinations should return departureList and arrivalList 1`] = `
Object {
  "arrivalList": Array [
    "Amsterdam",
    "Paris",
    "Warsaw",
    "Brussels",
    "London",
    "Stockholm",
    "Prague",
    "Moscow",
    "Madrid",
    "Geneva",
    "Budapest",
    "Kyiv",
    "Lisbon",
    "Rome",
    "Athens",
    "Istanbul",
  ],
  "departureList": Array [
    "London",
    "Amsterdam",
    "Warsaw",
    "Stockholm",
    "Paris",
    "Brussels",
    "Prague",
    "Moscow",
    "Madrid",
    "Geneva",
    "Budapest",
    "Kyiv",
    "Lisbon",
    "Rome",
    "Athens",
    "Istanbul",
  ],
}
`;
